File: mobile/ios/App/App/scout/LiteRtScoutGemmaEngine.swift
import Foundation

#if canImport(LiteRTLM)
import LiteRTLM

/// Independently resolves a native generation result exactly once. The LiteRT
/// stream may fail to finish even after cancellation; this gate lets the 120s
/// watchdog release the Capacitor call without waiting on that stuck stream.
private final class ScoutGemmaResultGate: @unchecked Sendable {
    private let lock = NSLock()
    private var continuation: CheckedContinuation<GenerateResult, Error>?
    private var result: Result<GenerateResult, Error>?
    private var tasks: [Task<Void, Never>] = []

    func install(_ continuation: CheckedContinuation<GenerateResult, Error>) {
        lock.lock()
        if let result = result {
            lock.unlock()
            continuation.resume(with: result)
            return
        }
        self.continuation = continuation
        lock.unlock()
    }

    func attach(_ tasks: Task<Void, Never>...) {
        lock.lock()
        if result != nil {
            lock.unlock()
            tasks.forEach { $0.cancel() }
            return
        }
        self.tasks.append(contentsOf: tasks)
        lock.unlock()
    }

    func resolve(_ result: Result<GenerateResult, Error>) {
        lock.lock()
        guard self.result == nil else {
            lock.unlock()
            return
        }
        self.result = result
        let continuation = self.continuation
        self.continuation = nil
        let tasks = self.tasks
        self.tasks.removeAll()
        lock.unlock()

        tasks.forEach { $0.cancel() }
        continuation?.resume(with: result)
    }
}

/// On-device Gemma 4 engine backed by LiteRT-LM's Swift API. Inference runs
/// entirely on the phone — no paid/cloud model APIs. The Swift mirror of
/// Android's `LiteRtScoutGemmaEngine`, against the same `.litertlm` model file
/// and the same `ScoutGemmaEngine` contract.
///
/// Compiled ONLY when the LiteRT-LM Swift package is present (`canImport`), so
/// the app builds green before the package is added. Verify these calls against
/// the installed package version when activating — the API is Early Preview
/// (see docs/runbooks/ios-scout-gemma-bridge.md).
///
/// iOS 15.0 minimum: LiteRT-LM Swift requires iOS 15; the project deployment
/// target is 14.0, so the class and its creation path are guarded with
/// @available / #available so iOS 14 devices cleanly fall back to the
/// UnavailableScoutGemmaEngine stub without a runtime crash.
@available(iOS 15.0, *)
final class LiteRtScoutGemmaEngine: ScoutGemmaEngine {
    private static let runtimeMaxNumTokens = 32_768
    private static let maxOutputTokensPerTurn = 2_048
    private static let generationTimeoutSeconds = 120

    private let modelPath: String
    private let cacheDir: String
    private let info: ScoutGemmaModelInfo
    private var engine: Engine?
    private var runtimePoisonedReason: String?

    private init(modelPath: String, cacheDir: String, info: ScoutGemmaModelInfo) {
        self.modelPath = modelPath
        self.cacheDir = cacheDir
        self.info = info
    }

    var isAvailable: Bool { true }

    func describeModel() -> ScoutGemmaModelInfo? { info }

    func warmUp() async throws {
        _ = try await ensureEngine()
    }

    func generate(prompt: String, systemContext: String, maxTokens: Int) async throws -> GenerateResult {
        do {
            let engine = try await ensureEngine()
            let outputTokenLimit = min(max(maxTokens, 1), Self.maxOutputTokensPerTurn)
            let trimmedSystemContext = systemContext.trimmingCharacters(in: .whitespacesAndNewlines)
            let conversationConfig = ConversationConfig(
                systemMessage: trimmedSystemContext.isEmpty
                    ? nil
                    : Message(trimmedSystemContext, role: .system),
                maxOutputTokens: outputTokenLimit)
            let conversation = try await engine.createConversation(with: conversationConfig)

			return try await generateBounded(
				conversation: conversation,
				prompt: prompt)
        } catch let honest as ScoutGemmaError {
            if case .timedOut = honest {
                runtimePoisonedReason =
                    "The previous Gemma turn timed out. Close and reopen Scout before trying local AI again."
            }
            throw honest
        } catch {
            // Never fabricate output: any runtime failure surfaces as unavailable.
            throw ScoutGemmaError.unavailable("On-device Gemma generation failed: \(Self.describeNativeError(error))")
        }
    }

    /// Runs one streamed turn with two independent native boundaries:
    /// ConversationConfig enforces the real LiteRT-LM decode-token limit, while
    /// Conversation.cancel() interrupts a wedged turn after the timeout or when
    /// the parent Swift task is cancelled. The plugin serializes calls into this
    /// method so LiteRT-LM never initializes or infers two turns at once.
	private func generateBounded(
		conversation: Conversation,
		prompt: String
	) async throws -> GenerateResult {
        let gate = ScoutGemmaResultGate()
        return try await withTaskCancellationHandler(
            operation: {
                try await withCheckedThrowingContinuation { continuation in
                    gate.install(continuation)

                    let generationTask = Task<Void, Never> {
                        do {
                            var text = ""
                            for try await chunk in conversation.sendMessageStream(
                                Message(prompt, role: .user)) {
                                text.append(chunk.toString)
                            }
                            if Task.isCancelled {
                                gate.resolve(.failure(ScoutGemmaError.cancelled))
                            } else {
                                // The native session enforces the exact decode cap,
                                // but its Swift stream exposes no finish reason.
                                gate.resolve(.success(GenerateResult(text: text, truncated: false)))
                            }
                        } catch {
                            gate.resolve(
                                .failure(Task.isCancelled ? ScoutGemmaError.cancelled : error))
                        }
                    }

                    let watchdogTask = Task<Void, Never> {
                        do {
                            try await Task.sleep(
                                nanoseconds: UInt64(Self.generationTimeoutSeconds) * 1_000_000_000)
                        } catch {
                            return
                        }
                        guard !Task.isCancelled else { return }
                        // Resolve first: even if the native cancel call wedges, the
                        // plugin lane and UI are released at the hard deadline.
                        gate.resolve(.failure(
                            ScoutGemmaError.timedOut(seconds: Self.generationTimeoutSeconds)))
                        try? conversation.cancel()
                    }
                    gate.attach(generationTask, watchdogTask)
                }
            },
            onCancel: {
                gate.resolve(.failure(ScoutGemmaError.cancelled))
                try? conversation.cancel()
            })
    }

    /// `initialize()` is heavy (seconds); run it lazily on the first generate.
    private func ensureEngine() async throws -> Engine {
        if let reason = runtimePoisonedReason {
            throw ScoutGemmaError.unavailable(reason)
        }
        if let engine = engine { return engine }
        // LiteRT-LM's default Gemma 4 E2B cache is 4K on this export. Request a
        // larger but phone-testable cache so Scout's long-context JS budget is
        // backed by the native runtime instead of only the model descriptor.
        let maxNumTokens = min(info.maxContextTokens, Self.runtimeMaxNumTokens)
        let config = try EngineConfig(
            modelPath: modelPath,
            backend: .cpu(),
            maxNumTokens: maxNumTokens,
            cacheDir: cacheDir)
        // Verified against the real LiteRT-LM 0.13.x Swift API: the initializer
        // takes the `engineConfig:` argument label (compile-checked via a scratch
        // SwiftPM target — see docs/runbooks/ios-scout-gemma-bridge.md).
        let created = Engine(engineConfig: config)
        try await created.initialize()
        engine = created
        return created
    }

    private static func describeNativeError(_ error: Error) -> String {
        let nsError = error as NSError
        if nsError.domain.isEmpty {
            return error.localizedDescription
        }
        return "\(error.localizedDescription) [domain=\(nsError.domain) code=\(nsError.code)]"
    }

    /// Returns the engine only when a checksum-verified model is on the device;
    /// nil otherwise so the plugin falls back to the fail-closed stub.
    static func tryCreate(store: ScoutModelStore) -> ScoutGemmaEngine? {
        guard store.status().state == ScoutModelStatus.ready else { return nil }
        let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path
            ?? NSTemporaryDirectory()
        return LiteRtScoutGemmaEngine(
            modelPath: store.modelFileURL.path,
            cacheDir: cacheDir,
            info: UnavailableScoutGemmaEngine.targetModel
        )
    }
}
#endif

/// Builds the active engine. Tries the LiteRT-LM engine first when its package is
/// present and a verified model exists, otherwise returns the fail-closed stub —
/// so the plugin compiles and runs identically with or without the package.
/// The LiteRT branch is further guarded with `#available(iOS 15.0, *)` because
/// LiteRT-LM requires iOS 15 while the project deployment target is 14.0.
enum ScoutGemmaEngineFactory {
    static func create(store: ScoutModelStore) -> ScoutGemmaEngine {
        #if canImport(LiteRTLM)
        if #available(iOS 15.0, *) {
            if let engine = LiteRtScoutGemmaEngine.tryCreate(store: store) {
                return engine
            }
        }
        #endif
        return UnavailableScoutGemmaEngine()
    }
}

File: mobile/android/app/src/main/java/com/hoggcountry/trailassistant/scout/LiteRtScoutGemmaEngine.kt
package com.hoggcountry.trailassistant.scout

import android.content.Context
import android.util.Log
import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Content
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.Message
import com.google.ai.edge.litertlm.MessageCallback
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * On-device Gemma 4 engine backed by LiteRT-LM
 * ([com.google.ai.edge.litertlm]), Google's cross-platform on-device LLM
 * runtime. Inference runs entirely on the phone — no paid/cloud model APIs.
 *
 * This is the single chosen runtime for the unified (Android + future iOS)
 * Scout engine; the iOS Swift plugin mirrors this same `ScoutGemmaEngine`
 * contract against the LiteRT-LM Swift API and the same `.litertlm` model file.
 *
 * Fail-closed by construction. [tryCreate] returns `null` (never throws) unless
 * [ScoutModelStore] reports a checksum-verified model on the device, so the
 * plugin falls back to [UnavailableScoutGemmaEngine] and Gemma-only builds block
 * chat instead of fabricating output or drifting to a cloud model. The heavy
 * [Engine.initialize] call (seconds) is deferred to the first [generate] so app
 * start is never blocked; [ScoutGemmaPlugin] already runs generation off the
 * main thread.
 */
class LiteRtScoutGemmaEngine private constructor(
    private val modelPath: String,
    private val cacheDir: String,
    private val modelInfo: ScoutGemmaModelInfo
) : ScoutGemmaEngine {

    // Guarded by synchronized(this) for both lazy init and close().
    private var engine: Engine? = null

    override fun isAvailable(): Boolean = true

    override fun describeModel(): ScoutGemmaModelInfo = modelInfo

    /**
     * Eagerly run the costly [Engine.initialize] so the first real chat turn is
     * fast and doesn't risk the cold GPU/CPU-delegate init failing under the
     * user's first message. Best-effort and never throws — if warm-up fails the
     * engine stays null and the next [generate] retries init exactly as before.
     */
    override fun warmUp() {
        try {
            ensureEngine()
        } catch (failure: Throwable) {
            Log.w(TAG, "Engine warm-up failed (will retry on first generate): ${failure.message}")
        }
    }

    @Throws(ScoutGemmaUnavailableException::class)
    override fun generate(
        prompt: String,
        systemContext: String?,
        maxTokens: Int,
        sink: ScoutGemmaEngine.TokenSink?
    ): ScoutGemmaEngine.GenerateResult {
        val input = if (systemContext.isNullOrEmpty()) prompt else "$systemContext\n\n$prompt"
        try {
            val active = ensureEngine()
            // A fresh conversation per call keeps turns independent and releases
            // native resources deterministically (Conversation is Closeable).
            active.createConversation().use { conversation ->
                val full = StringBuilder()
                val failure = arrayOfNulls<Throwable>(1)
                val done = CountDownLatch(1)
                // Stream tokens as they are produced. onMessage delivers
                // incremental Content parts; we forward each text chunk to the
                // sink and accumulate the full answer to return.
                conversation.sendMessageAsync(
                    input,
                    object : MessageCallback {
                        override fun onMessage(message: Message) {
                            val chunk = message.contents.contents
                                .filterIsInstance<Content.Text>()
                                .joinToString("") { it.text }
                            if (chunk.isNotEmpty()) {
                                full.append(chunk)
                                sink?.onToken(chunk)
                            }
                        }

                        override fun onDone() {
                            done.countDown()
                        }

                        override fun onError(t: Throwable) {
                            failure[0] = t
                            done.countDown()
                        }
                    },
                    emptyMap<String, Any>()
                )
                // Bounded wait: if the native runtime wedges and never calls
                // onDone/onError, fail honestly instead of blocking the single
                // inference thread (and every later turn) forever. Exiting the
                // `use` block closes the Conversation, aborting the in-flight call.
                val completed = done.await(GENERATE_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                if (!completed) {
                    throw ScoutGemmaUnavailableException(
                        "On-device Gemma timed out after ${GENERATE_TIMEOUT_SECONDS}s — give it a moment and try again."
                    )
                }
                failure[0]?.let { throw it }

                val text = full.toString()
                if (text.isBlank()) {
                    // Diagnostic: engine ran but produced no Content.Text.
                    // `adb logcat -s ScoutGemma:*` surfaces this.
                    Log.w(TAG, "Gemma stream produced no text parts")
                }
                return ScoutGemmaEngine.GenerateResult(text, false)
            }
        } catch (alreadyHonest: ScoutGemmaUnavailableException) {
            // Preserve the specific message (e.g. timeout) — don't re-wrap it.
            throw alreadyHonest
        } catch (oom: OutOfMemoryError) {
            // The model needs substantial RAM. Surface an honest device-capability
            // message and drop the engine so a retry starts clean. Never fabricate.
            Log.w(TAG, "Out of memory running on-device model", oom)
            close()
            throw ScoutGemmaUnavailableException(
                "Not enough free memory to run the on-device model on this device."
            )
        } catch (failure: Throwable) {
            // Never fabricate output: any runtime failure surfaces as unavailable.
            throw ScoutGemmaUnavailableException(
                "On-device Gemma generation failed: ${failure.message}"
            )
        }
    }

    /**
     * Releases the native [Engine] if one has been initialized. Safe to call
     * multiple times; subsequent calls are no-ops. Called by [ScoutGemmaPlugin]
     * when swapping engines or when the plugin is destroyed.
     */
    override fun close() {
        synchronized(this) {
            engine?.close()
            engine = null
        }
    }

    @Synchronized
    private fun ensureEngine(): Engine {
        engine?.let { return it }
        // Auto-select the fastest backend the device supports, silently: try the
        // GPU accelerator first (far faster than CPU on capable chips), fall back
        // to CPU if the GPU delegate can't initialize. Free (no bigger download),
        // so it is never a user choice in the default flow; an advanced override
        // can force a specific backend later.
        var lastError: Throwable? = null
        for (textBackend in listOf(Backend.GPU(), Backend.CPU())) {
            try {
                Log.i(TAG, "Initializing LiteRT-LM engine on ${textBackend.name} backend (can take seconds)")
                // Only the text backend is accelerated. The model constrains its
                // audio backend to CPU (and we use no audio/vision in Scout chat),
                // so forcing GPU on those slots fails engine creation outright.
                val config = EngineConfig(
                    modelPath = modelPath,
                    backend = textBackend,
                    visionBackend = Backend.CPU(),
                    audioBackend = Backend.CPU(),
                    maxNumTokens = minOf(modelInfo.maxContextTokens, RUNTIME_MAX_NUM_TOKENS),
                    maxNumImages = null,
                    cacheDir = cacheDir
                )
                val created = Engine(config)
                created.initialize()
                Log.i(TAG, "LiteRT-LM engine initialized on ${textBackend.name}")
                engine = created
                return created
            } catch (failure: Throwable) {
                lastError = failure
                Log.w(TAG, "Backend ${textBackend.name} unavailable (${failure.message}); trying next.")
            }
        }
        throw lastError ?: IllegalStateException("No usable LiteRT-LM backend available")
    }

    companion object {
        private const val TAG = "ScoutGemma"
        private const val RUNTIME_MAX_NUM_TOKENS = 32_768

        // Upper bound for a single on-device generation. Generous enough for a
        // long answer on a slow device, but bounded so a wedged native callback
        // can't hang the inference thread (and all later turns) indefinitely.
        private const val GENERATE_TIMEOUT_SECONDS = 120L

        /**
         * Loads the LiteRT-LM engine against the verified on-device model, or
         * returns `null` (never throws) so the plugin can fall back to the stub.
         * Cheap: only inspects model status and constructs the engine wrapper —
         * the costly [Engine.initialize] runs lazily on first generate.
         */
        @JvmStatic
        fun tryCreate(context: Context): ScoutGemmaEngine? {
            return try {
                val store = ScoutModelStore(context)
                val status = store.getStatus()
                if (ScoutModelStatus.READY != status.state) {
                    // Fail-closed: never load an unverified or absent model file.
                    return null
                }
                val modelFile = store.modelFile()
                if (!modelFile.isFile) {
                    return null
                }
                LiteRtScoutGemmaEngine(
                    modelFile.absolutePath,
                    context.cacheDir.absolutePath,
                    UnavailableScoutGemmaEngine.TARGET_MODEL
                )
            } catch (failure: Throwable) {
                Log.w(TAG, "LiteRT-LM Gemma engine unavailable: ${failure.message}")
                null
            }
        }
    }
}
